File: s_coding/삼성 SW 역량 테스트 기출 문제/15686_치킨배달.cpp
#if 0
//치킨배달
//https://www.acmicpc.net/problem/15686
#include <iostream>
#include <algorithm>
#include <vector>
#define MAX 50
using namespace std;

int map[MAX][MAX];
int N, M;
int sol = 0x7fffffff;

struct pos {
	int h;
	int w;
};
vector<pos> home, chicken,selected;


void DFS(int idx,int cnt ) { //조합 재귀 

	//if (n > chicken.size())return; //종료조건 : 치킨집 개수보다 클 때
	if (cnt ==	M) { 
		int sum=0;

		for (int i = 0; i < (int)home.size(); i++) {
			int d = 99999999;
		
			for (int j = 0; j < (int)selected.size(); j++) {
			//for (auto j : selected) {
				d = min(d, abs(home[i].h - selected[j].h) + abs(home[i].w - selected[j].w));
			}
			sum += d;

		}
		sol = min(sol, sum);
		return;
	}

	for (int i = idx; i < (int)chicken.size(); i++) {
		//치킨집 선택 
		selected.push_back(chicken[i]);
		DFS(i + 1, cnt + 1);
		selected.pop_back();
	}


}
void InputData() {
	
	scanf("%d %d", &N, &M);
	for (int i = 0; i < N; i++) {
		for (int j = 0; j < N; j++) {
			int tmp;
			scanf("%d", &tmp);
			if (tmp == 1) home.push_back({ i,j }); //집
			else if (tmp == 2) chicken.push_back({ i,j }); //치킨
		}
	}


}
int main() {

	freopen("input.txt", "r", stdin);

	InputData();
	DFS(0,0);
	printf("%d", sol);

}
#endif

File: s_coding/삼성 SW 역량 테스트 기출 문제/15686-치킨배달.py
#치킨배달
#https://www.acmicpc.net/problem/15686
import sys
sys.stdin = open("input.txt","r")
n,m=map(int,input().split())
board=[list(map(int,input().split())) for _ in range(n)]
home=[]
chicken=[] #치킨집
result=[]
for i in range(n):
    for j in range(n):
        if board[i][j]==1:#집
            home.append((i,j))
        if board[i][j]==2: #치킨집
            chicken.append((i,j))
check=[0]*(len(chicken)) #치킨집 체크 배열
selected=[] #선택된 치킨집의 개수
def chicken_distance(chick):
    dist = [5000] * (len(home))
    for i,j in chick:
        for k in range(len(home)):
            a,b=home[k]
            tmp=abs(i-a)+abs(j-b)
            dist[k]=min(dist[k],tmp)
    return sum(dist)

#치킨집 중에서 m개 만큼 뽑음
def dfs(d,s):
    res=0
    if d==m:
        for i in range(len(chicken)):
            if check[i]==1:
                selected.append(chicken[i])
        #print(selected)
        cd=chicken_distance(selected)
        res=max(cd,res)
        selected.clear()
        #print(res)
        result.append(res)
    else:
        for i in range(s,len(chicken)):
            check[i]=1
            dfs(d+1,i+1)
            check[i]=0


dfs(0,0)
print(min(result))
